Packages/vcs/Lib/configurator.py

```diff
@@ -258,81 +199,124 @@ def detach(self):
-            if clicked:
-                if self.target is None or self.target.is_object(clicked) == False:
-                    self.activate(clicked, display_clicked)
-                elif self.shift() and self.target.contains_object(clicked):
-                    self.target.remove_object(clicked)
-                    if len(self.target.targets) == 0:
-                        self.target.detach()
-                        self.target = None
-                        self.save()
-                        return
-            elif self.target is not None and self.shift() == False:
+            clicked_actor = self.actor_at_point(*point)
+
+            if clicked_actor is None and self.target:
                 self.deactivate(self.target)
                 return
 
+            if clicked_actor is None:
+                return
+
+            display, key = self.display_and_key_for_actor(clicked_actor)
+            if editable_type(display, key):
+                # Some methods (markers) have more than one actor per displayed item
+                if clicked_actor != display.backend[key] and clicked_actor not in display.backend[key]:
+                    for group in display.backend[key]:
+                        if clicked_actor in group:
+                            clicked_actor = group
+                            break
+
+                self.activate(display, clicked_actor, key)
+
         self.clicking = None
 
+    def display_and_key_for_actor(self, actor):
+        """
+        Iterates across the displays, and tries to find the actor in the backend.
+        """
+        for display in self.displays:
+            for key in display.backend:
+                try:
+                    if actor == display.backend[key] or actor in display.backend[key] or [True for group in display.backend[key] if actor in group]:
+                        return display, key
+                except TypeError:
+                    # display.backend[key] isn't iterable
+                    pass
+            else:
+                continue
+            break
+        else:
+            return None, None
+
+
+    def actor_at_point(self, x, y):
+        """
+        Iterates all renderers, checks if there's an actor at the point
+        """
+        obj = None
+        layer_obj = 0
+
+        for ren in vtkIterate(self.render_window.GetRenderers()):
+            layer = ren.GetLayer()
+
+            if self.interactor is not None:
+                manager = vcs.vtk_ui.manager.get_manager(self.interactor)
+                if ren == manager.actor_renderer or ren == manager.renderer:
+                    continue
+
+            if obj is not None and layer < layer_obj:
+                continue
+
+            if self.picker.PickProp(x, y, ren):
+                obj = self.picker.GetViewProp()
+                layer_obj = layer
+
+        return obj
+
+
     def hover(self, object,event):
         if self.clicking is not None:
             return
```

Comment:
Heh, never heard of that one before: http://en.wikipedia.org/wiki/Don%27t_repeat_yourself The backronyms for WET are pretty enjoyable, too.